### PR TITLE
User filter update to support enterprise domains for User History Report

### PR DESCRIPTION
**Pull request by stephherbers**
## Product Description

[ticket](https://dimagi-dev.atlassian.net/browse/USH-1316). This is "stage 1" of the changes to the user filter where we are just isolating this change to the User History report, and not other reports. The next stage will likely be having this available for all reports for enterprise domains, but that ticket has yet to be prioritized. 

## Technical Summary

Started with refactoring.
Then for the correct options to appear, it will check for `user_history` in the request.
And for the results to show, the optional param is added to the `user_es_query` and is only set to `True` for the user history report
The functionality for `get_enterprise_domains` was already in place before this PR and returns the source domain if a mirrored domain. The new functionality of `get_enterprise_controlled_domains` returns all the mirrored domains if a source domain is inputted.

## Feature Flag

[USER_HISTORY_REPORT](https://confluence.dimagi.com/pages/viewpage.action?spaceKey=saas&title=User+History+Report) and [DOMAIN_PERMISSIONS_MIRROR](https://confluence.dimagi.com/display/saas/Enterprise+Permissions)

### Safety story

full functionality was testing on staging for enterprise and non-enterprise domains and also within the source and mirrored domains. I confirmed no change for the user filter in other reports, and the correct information for the mirrored domains.

### Automated test coverage

Added tests in `test_enterprise_utils` and modified in `test_esaccessors.py`

### QA Plan

I am not requesting QA

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by stephherbers** [comment]
hey @orangejenny @AddisonDunn tagging you here

File: corehq/apps/es/users.py [hunk 1]
```diff
@@ -64,14 +64,8 @@ def show_only_inactive(self):
...
 
-def domain(domain, allow_enterprise=False):
-    domain_list = [domain]
-    if allow_enterprise:
-        from corehq.apps.enterprise.models import EnterprisePermissions
-        config = EnterprisePermissions.get_by_domain(domain)
-        if config.is_enabled and domain in config.domains:
-            domain_list.append(config.source_domain)
-    return domains(domain_list)
+def domain(domain):
```
Comment (orangejenny) [inline corehq/apps/es/users.py:67]: It really is nice to get the enterprise logic out of here and into the `enterprise` app - thank you for cleaning this up.

File: corehq/apps/reports/filters/controllers.py [hunk 2]
```diff
@@ -122,10 +123,19 @@ def _get_users(self, query, start, size, include_inactive=False):
         users = customize_user_query(self.request.couch_user, self.domain, users)
         return [self.utils.user_tuple(u) for u in users.run().hits]
 
+    def _should_include_enterprise_controlled_domains(self):
+        from corehq.apps.reports.standard.users.reports import UserHistoryReport
+        if UserHistoryReport.slug in self.request.META['HTTP_REFERER']:
```
Comment (orangejenny) [inline corehq/apps/reports/filters/controllers.py]: Nitpick: you could do f"/{UserHistoryReport.slug}/", even though the slug is unlikely to coincidentally appear elsewhere in the URL.

**Review by orangejenny — COMMENTED** [review]
Question betraying my unfamiliarity with reports: why do you need to check the referer and also set the new flag in `UserHistoryReport`? I'd have thought you'd need to just do one of those things.

**Comment by stephherbers** [comment]
Hey @orangejenny I found out that in order for the users from the enterprise domains to appear in the filter options, the change is needed in `corehq/apps/reports/filters/controllers.py`. And for the results to appear after hitting "Apply", the flag needs to be added to `UserHistoryReport` and `user_es_query`. 
I was a bit surprised given that we already had the functionality of`get_enterprise_domains` in `user_es_query`, but then noticed that it just appears in the results and you cannot select users from the other domains returned from that function.
For this ticket, I interpreted that we would want to be able to select those users as well to filter-- curious if you viewed it the same way.

**Comment by orangejenny** [comment]
Ohhh, that makes sense, one change for the user options and one for the results.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/es/users.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/es/users.py:67]: :star2:

**Comment by stephherbers** [comment]
ready to merge but waiting to discuss approach with @snopoke

**Comment by snopoke** [comment]
I don't think we've thought this through very well. The current filters allow filtering on groups and locations as well as just usernames. Is the intention that the filter include groups and locations from ALL domains in the enterprise? I don't think the user experience for that will be great and the performance may also not be good. Further more it would require significantly more changes to the filters.

For the user history report specifically I don't think the group / location filters are that useful which is another reason why I removed those options in #30854 and only only allowed selecting / searching users.

I also concerned about the security with these changes. Since they are applied to ALL user filters any errors could expose users from one domain to another. I think it would be better to take a very targeted approach rather than changing the base filter directly.

**Comment by mkangia** [comment]
> For the user history report specifically I don't think the group / location filters are that useful 

Just adding some context that the filters were added just as what I thought would be useful and were not necessarily requirements. So I think we can still modify these filters based on what feels best for the report overall and limit them. We might need to check with other projects (the ones using that feature flag to access report) that are also looking for this report though.

**Comment by snopoke** [comment]
> As we expand the enterprise domains to other reports 

Is there a ticket / doc for this requirement? I'm concerned about creating cross domain data access without fully considering the implications.

>Regarding the security concerns, I have tested specifically for these concerns in mind

Were you able to test whether a user can spoof the referrer to trigger the enterprise filtering on reports that it shouldn't be available on e.g. case list? Since `HTTP_REFERRER` is just a header it's very easy to change it to anything you like which makes it a poor candidate for logic decisions.

**Pull request merged**